src/idpi/operators/curl.py

```diff
@@ -0,0 +1,122 @@
+import xarray as xr
+import numpy as np
+
+from .. import constants as const
+
+DEFAULT_DIFF_TYPE = "center"
+STENCILS = {
+    "left": {"m": slice(0, -2), "c": slice(1, -1), "p": slice(1, -1)},
+    "center": {"m": slice(0, -2), "c": slice(1, -1), "p": slice(2, None)},
+    "right": {"m": slice(1, -1), "c": slice(1, -1), "p": slice(2, None)},
+}
+
+
+def stpt(stencil: str, diff_type: str | None = None) -> dict[str, slice]:
+    """
+    Return a dictionary which indicates the stencil points of a data array as specified by the argument.
+
+    This dictionary can be used for selecting values from a data array.
+    Consider the following example, where ``A`` is a data array. ::
+
+        B = A[stpt('ccc')] + A[stpt('cpm')]
+
+    Here we add for every point a in ``A`` its upper (y+1) neighbor, one level below (z-1) and store the result in ``B``.
+    In this method we allow 3d neighbor stencils. Points on the border won't be included.
+
+    Depending on the ``diff_type`` (set with ``set_diff_type``), the stencils will be interpreted differently.
+    For 'left', positive stencils (+1, 'p') will be mapped to the original point.
+    For 'right', negative stencils (-1, 'm') will be mapped to the original point.
+
+    Args:
+        stencil (str): A string indicating the stencil.
+        This string must have three characters, corresponding to the 'x', 'y' and 'z' dimension (in this order).
+        The characters can be
+
+        - 'c': center (+/-0)
+        - 'm': below (-1)
+        - 'p': above (+1)
+    """
+    if diff_type is None:
+        diff_type = DEFAULT_DIFF_TYPE
+    s = STENCILS[diff_type]
+    return {d: s[k] for d, k in zip("xyz", stencil)}
+
+
+class FDField:
+    def __init__(self, field: xr.DataArray, diff_type: str = DEFAULT_DIFF_TYPE):
+        self.field = field
+        self.diff_type = diff_type
+
+    def __getitem__(self, stencil: str):
+        return self.field[stpt(stencil, diff_type=self.diff_type)]
+
+    @classmethod
+    def from_fields(cls, *fields, diff_type: str = DEFAULT_DIFF_TYPE):
+        for field in fields:
+            yield cls(field, diff_type)
+
+
+def curl(
+    U: xr.DataArray,
+    V: xr.DataArray,
+    W: xr.DataArray,
+    HHL: xr.DataArray,
+    inv_dlon: float,
+    inv_dlat: float,
+    wk: float,
+    sqrtg_r_s: xr.DataArray,
+    lat: xr.DataArray,
+) -> tuple[xr.DataArray, xr.DataArray, xr.DataArray]:
+    # prepare parameters
+
+    r_earth_inv = 1 / const.earth_radius
+    deg2rad = np.pi / 180
+    acrlat: xr.DataArray = 1 / (np.cos(lat * deg2rad) * const.earth_radius)  # type: ignore[assignment]
+    tgrlat: xr.DataArray = np.tan(lat * deg2rad)  # type: ignore[assignment]
```

Comment:
One of the issues in the original branch is that the lattitude values are in degrees and not micro degrees.